### PR TITLE
[orchestrator] Collect Ingress resources

### PR DESCRIPTION
**Pull request by ahmed-mez**
### What does this PR do?

Collect `Ingress` resources in the orchestrator check.

### Additional Notes

- Added `IsAvailable()` method to the collector interface, no-op for the existing collectors
- The Ingress collector declares itself not available if the `networking.k8s.io/v1` api is not available (k8s < 1.19)

### Describe how to test/QA your changes

- To be tested on k8s < 1.19 and >= 1.19
- On k8s < 1.19 the ingress collector shouldn't start and shouldn't impact the check and the other collectors

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -429,3 +432,5 @@ replace (
 	github.com/go-logr/logr => github.com/go-logr/logr v0.4.0
 	github.com/go-logr/stdr => github.com/go-logr/stdr v0.4.0
 )
+
+replace github.com/DataDog/agent-payload/v5 v5.0.12 => github.com/DataDog/agent-payload/v5 v5.0.14-0.20220208101347-d5c0c45a5924
```
Comment (ahmed-mez) [inline go.mod:436]: this is temporary

**Review by alai97 — APPROVED** [review]
Looks good for docs!

File: pkg/collector/corechecks/cluster/orchestrator/transformers/k8s/ingress_test.go [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+func TestExtractIngressStatus(t *testing.T) {
+	errMsg := "error"
+	tests := map[string]struct {
+		input    netv1.IngressStatus
+		expected model.IngressStatus
+	}{
+		"empty": {input: netv1.IngressStatus{}, expected: model.IngressStatus{Ingress: []*model.LoadBalancerIngress{}}},
+		"multiple ingress status": {
```
Comment (xlucas) [inline pkg/collector/corechecks/cluster/orchestrator/transformers/k8s/ingress_test.go:144]:
```suggestion
		"multiple ingress statuses": {
```

File: pkg/collector/corechecks/cluster/orchestrator/transformers/k8s/ingress_test.go [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/core/v1"
+	netv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestExtractIngress(t *testing.T) {
+	pathType := netv1.PathTypeImplementationSpecific
+	apiGroup := "apiGroup"
+	ingressClassName := "ingressClassName"
```
Comment (xlucas) [inline pkg/collector/corechecks/cluster/orchestrator/transformers/k8s/ingress_test.go:25]: tip 💡: you can use this test helper to build string pointers directly https://github.com/DataDog/datadog-agent/blob/96fea554086547328f93c41d63746aa20ce461ec/pkg/collector/corechecks/cluster/orchestrator/transformers/k8s/common_test.go#L23-L25

File: pkg/collector/corechecks/cluster/orchestrator/transformers/k8s/ingress_test.go [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+			},
+			expected: model.IngressBackend{
+				Resource: &model.TypedLocalObjectReference{
+					ApiGroup: "apiGroup",
+					Kind:     "kind",
+					Name:     "name",
+				},
+			},
+		},
+		"with service": {
```
Comment (xlucas) [inline pkg/collector/corechecks/cluster/orchestrator/transformers/k8s/ingress_test.go:251]: IIUC from the proto definition service name and port are mutually exclusive, do you think it'd make sense to split this in two to be more "realistic"?

**Review by xlucas — COMMENTED** [review]
Great job!

The main thing I've in mind is whether we could improve the discovery of Kubernetes API groups available in the cluster,  the rest is pretty minor.

File: pkg/collector/corechecks/cluster/orchestrator/transformers/k8s/ingress_test.go (hunk 4 above)
Comment (ahmed-mez) [reply pkg/collector/corechecks/cluster/orchestrator/transformers/k8s/ingress_test.go:248]: Actually, service and reference are mutually exclusive in `IngressBackend`. But a service and port can coexist in `IngressServiceBackend`. (I hope I understood the question correctly)

File: pkg/collector/corechecks/cluster/orchestrator/collector_bundle.go [hunk 5]
```diff
@@ -114,6 +114,11 @@ func (cb *CollectorBundle) Initialize() error {
 
 	for _, collector := range cb.collectors {
 		collector.Init(cb.runCfg)
+		if available := collector.IsAvailable(); !available {
```
[the remark this answers is not in the document]
Comment (ahmed-mez) [reply pkg/collector/corechecks/cluster/orchestrator/collector_bundle.go]: Good points thanks!

The collector would run indeed but doesn't log any errors just empty results because the informer isn't running. Added some logic to remove unavailable collectors 👍

File: pkg/collector/corechecks/cluster/orchestrator/collectors/k8s/ingress.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+	c.retryLister = rcfg.APIClient.Cl.NetworkingV1().Ingresses("").List
+}
+
+// IsAvailable returns whether the collector is available.
+// Returns false if the networking.k8s.io/v1 API version is not available (kubernetes < 1.19).
+func (c *IngressCollector) IsAvailable() bool {
+	var retrier retry.Retrier
+	if err := retrier.SetupRetrier(&retry.Config{
+		Name:          "NetworkV1Discovery",
+		AttemptMethod: c.list,
```
[the remark this answers is not in the document]
Comment (ahmed-mez) [reply pkg/collector/corechecks/cluster/orchestrator/collectors/k8s/ingress.go:71]: Yes ideally the discovery client can be used for this. We had exactly the same need for the admission controller, we relied on the discovery client but then we had to switch to this approach. See full details in this [PR](https://github.com/DataDog/datadog-agent/pull/10171).
TL;DR The discovery clients expects answers from all registered APIs including `external.metrics.k8s.io/v1beta1` which is the role of the DCA to answer it when the external metrics provider is enabled, at startup the DCA is not ready (expected) so every discovery client request would fail.
The current approach is not perfect indeed but it's robust, the expensive list operation would only be executed 3 times worst case.

[the remark this answers is not in the document]
Comment (xlucas) [reply pkg/collector/corechecks/cluster/orchestrator/collectors/k8s/ingress.go:71]: 😮 

I'm not 100% sure if it means that:
- the DCA can block all responses to `/apis` from being returned for 30s
or
- the API for `external.metrics.k8s.io/v1beta1` doesn't show up in the list for 30s

File: pkg/collector/corechecks/cluster/orchestrator/transformers/k8s/ingress_test.go (hunk 4 above)
Comment (xlucas) [reply pkg/collector/corechecks/cluster/orchestrator/transformers/k8s/ingress_test.go:248]: I'm referring to this https://github.com/kubernetes/api/blob/v0.21.5/networking/v1/generated.proto#L493-L503

File: pkg/collector/corechecks/cluster/orchestrator/collectors/k8s/ingress.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ahmed-mez) [reply pkg/collector/corechecks/cluster/orchestrator/collectors/k8s/ingress.go:71]: It's the former, `30s` is the best case (minimum time required to be come ready). During that time (DCA not ready) the discovery requests fail with this error `unable to retrieve the complete list of server APIs: external.metrics.k8s.io/v1beta1: the server is currently unable to handle the request`. It's a shame that the discovery client doesn't return a partial list of the available APIs in this case, it's all or nothing.

File: pkg/collector/corechecks/cluster/orchestrator/transformers/k8s/ingress_test.go (hunk 4 above)
Comment (ahmed-mez) [reply pkg/collector/corechecks/cluster/orchestrator/transformers/k8s/ingress_test.go]: addressed 👍

File: pkg/collector/corechecks/cluster/orchestrator/collectors/k8s/ingress.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (xlucas) [reply pkg/collector/corechecks/cluster/orchestrator/collectors/k8s/ingress.go:71]: Thanks for clarifying, then it's rather disappointing 😕

Not in the scope of that PR but maybe there's something we could do better in the external metrics provider to stop this from happening? I suppose we'll hit this issue again if we try to collect CRDs later.

Let's start with your approach for now 👍

[the remark this answers is not in the document]
Comment (ahmed-mez) [reply pkg/collector/corechecks/cluster/orchestrator/collectors/k8s/ingress.go:71]: > but maybe there's something we could do better in the external metrics provider to stop this from happening?

It's a tough one, not ready pods don't serve service requests be design in k8s, and I'm not sure we want to remove the readiness prove for the cluster agent, so no obvious solution unfortunately. I'll add something to the backlog though to investigate deeper and try to find other alternatives.

File: pkg/collector/corechecks/cluster/orchestrator/collector_bundle.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (xlucas) [reply pkg/collector/corechecks/cluster/orchestrator/collector_bundle.go]: Sorry to be picky on this but instead of doing in-place updates could you more simply work on a new slice of `availableCollectors` and assign this to cb.collectors after the loop? As it's not performance sensitive let's favor greater readability 🙂

[the remark this answers is not in the document]
Comment (ahmed-mez) [reply pkg/collector/corechecks/cluster/orchestrator/collector_bundle.go]: done!

**Review by xlucas — APPROVED** [review]
